### PR TITLE
daemon: selectively reboot based on diffs of applied MCs

### PR DESCRIPTION
**Pull request by yuqi-zhang**
Adds to/supercedes https://github.com/openshift/machine-config-operator/pull/2254

This is the overall PR for the reboot epic work, which aims to allow "rebootless updates" when certain changes happen.

Currently supported:
No action (drain, update files, uncordon)
 - ssh keys
 - pull secret

Reload crio
 - registries.conf changes

Reboot
 - all other scenarios

Also add corresponding unit and e2e tests

### CODE REVIEW

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:485]: This should run only when file was found in both old and new MC.

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:581]: We are doing multiple things when we call updateFiles() irrespective of what rebootAction we perform like deleting stale file, writing units from newIgnConfig, etc. There are chances of getting errors in these steps. Cordon and drain will ensure that we don't end up with degraded node that has running workload.

File: pkg/daemon/update.go
Comment (yuqi-zhang) [reply pkg/daemon/update.go:485]: ah good catch! this works as it today I think but will cause a duplicate add of a file. Will fix

File: pkg/daemon/update.go
Comment (yuqi-zhang) [reply pkg/daemon/update.go:581]: Hmm ok, I think my consideration was that drain on some workloads (e.g. CI) take quite awhile so it would be a nice value add. And if we were doing an update that does not require a reboot, it was unlikely to break the state of the actual node (at worst, it would e.g. fail a file write, but would not commit any changes to disk and the node itself should still be working even as we're degrading the pool).
For safety purposes though perhaps we should do this drain method for now, and consider improving in the future?

Comment (sinnykumari) [reply pkg/daemon/update.go:581]: You are right, even I get tempted to skip drain which will work fine during happy path. This is probably an extra precaution to avoid coroner cases bugs that bite us later on.
+1 on revisiting drain optimization later on (probably in 4.8?)

**Comment by sinnykumari** [comment]
Tested all scenarios we are handling in this PR including node scaleup, working perfectly fine.

**Comment by yuqi-zhang** [comment]
Fixed above error case and added some initial unit tests. Will expand.

Manual testing looks good so far.

**Comment by sinnykumari** [comment]
Added e2e test

File: pkg/daemon/update.go
Comment (cgwalters) [inline pkg/daemon/update.go:54]: What change would result in no action?

File: pkg/daemon/update.go
Comment (cgwalters) [inline pkg/daemon/update.go:103]: One of the massive benefits of systemd over baseline Unix is that by using cgroups, it applies a rigorous *structure* to the prior total Wild West of Unix processes you could identify by grepping their name or maybe racy PID files.

In particular, this invocation will send SIGHUP to e.g. a process running in a regular user pod if it happens to be named "crio".

What you want instead is `systemctl reload crio` - systemd is actively tracking the PID that it launched and can send the signal in a race-free way to exactly the correct process.

File: pkg/daemon/update.go
Comment (yuqi-zhang) [reply pkg/daemon/update.go:54]: SSH keys + pull secret. I'll add some comments here.

"No action" is perhaps not the right way to frame it, we still cordon, drain + write files/units/etc. We just don't reload any unit and don't reboot.

File: pkg/daemon/update.go
Comment (yuqi-zhang) [reply pkg/daemon/update.go:103]: Ah for some reason I thought crio didn't have it defined. I can confirm that it indeed should be doing the same thing:

```
# systemctl cat crio.service
...
ExecReload=/bin/kill -s HUP $MAINPID
...
```

And running a reload:
```
Nov 30 23:23:17 ip-10-0-131-15 systemd[1]: Reloading Open Container Initiative Daemon.
Nov 30 23:23:17 ip-10-0-131-15 crio[1507]: time="2020-11-30 23:23:17.304012464Z" level=info msg="Reloading configuration"
Nov 30 23:23:17 ip-10-0-131-15 systemd[1]: Reloaded Open Container Initiative Daemon.
Nov 30 23:23:17 ip-10-0-131-15 crio[1507]: time="2020-11-30 23:23:17.309633738Z" level=info msg="Updating config from file /etc/crio/crio.conf.d/0-default"
Nov 30 23:23:17 ip-10-0-131-15 crio[1507]: time="2020-11-30 23:23:17.309845337Z" level=info msg="Updating config from path /etc/crio/crio.conf.d"
Nov 30 23:23:17 ip-10-0-131-15 crio[1507]: time="2020-11-30 23:23:17.310093463Z" level=info msg="Applied new registry configuration: &{Registries:] UnqualifiedSearchRegistries:[registry.access.redhat.com docker.io]}"
Nov 30 23:23:17 ip-10-0-131-15 crio[1507]: time="2020-11-30 23:23:17.310131904Z" level=info msg="No seccomp profile specified, using the internal efault"
Nov 30 23:23:17 ip-10-0-131-15 crio[1507]: time="2020-11-30 23:23:17.310144510Z" level=info msg="Set config seccomp_profile to \"\""
```

Seems like its doing what we want. I'll defer to @sinnykumari since she is more knowledgeable on whether that is correct.

**Comment by yuqi-zhang** [comment]
/retest

`uptimeOld, err := strconv.ParseFloat(oldTime, 64)` seems to have failed in the run but works locally, trying again just in case

**Comment by sinnykumari** [comment]
> /retest
> 
> `uptimeOld, err := strconv.ParseFloat(oldTime, 64)` seems to have failed in the run but works locally, trying again just in case

yeah, I don't know why it is failing in ci and works fine when running `make test-e2e` on a local cluster. This is making debugging difficult. From one of the ci run where I added debug message, it printed:
```
 mcd_test.go:193: DEBUG: Uptime file content: I1130 18:29:05.151521    8238 request.go:645] Throttling request took 1.019506334s, request: GET:https://api.ci-op-fpxq6893-1354f.origin-ci-int-gce.dev.openshift.com:6443/apis/storage.k8s.io/v1beta1?timeout=32s
```

File: pkg/daemon/update.go
Comment (sinnykumari) [reply pkg/daemon/update.go:103]: Ah, didn't know `crio reload` already does the same thing, definitely nicer way to go.
Useful learning and something I will keep in mind :)

File: test/e2e/mcd_test.go
Comment (kikisdeliveryservice) [inline test/e2e/mcd_test.go:165]: Looking at the test failure are you sure this oldTime is correct? maybe it should be a different index?

Comment (sinnykumari) [reply test/e2e/mcd_test.go:165]: command will return correct value, but it seems to go into api timeout  error when we call `cat /rootfs/uptime` through execCmdOnNode() . See the error message in https://github.com/openshift/machine-config-operator/pull/2259#issuecomment-736298449 from one of the previous gcp-op run on this [job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/2259/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1333453313349259264)

Comment (sinnykumari) [reply test/e2e/mcd_test.go:165]: While replying to you, I understand what is happening. execCmdOnNode() combines both stdout and error and returns it which is causing the problem.

**Comment by sinnykumari** [comment]
gcp-op test is green now :tada: 
/retest

**Comment by yuqi-zhang** [comment]
seems like a flake
/retest

**Comment by kikisdeliveryservice** [comment]
Going to pull up a cluster with this PR just to test it out :woman_dancing:

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:1130]: im a little confused about why we are returning bool here when we seem to just care about whether we return an error or not? where this is called we are always first checking error , so im not clear on why we need to return the bool bc it seems to always come down to: is there an error or not (and there's only one case where err is nil)

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [inline pkg/daemon/update.go:111]: wondering if there should be an event here?

File: test/e2e/mcd_test.go
Comment (kikisdeliveryservice) [inline test/e2e/mcd_test.go:157]: why break up the 2 kernel tests and put this in between?

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [inline pkg/daemon/update.go:122]: feel like this would make more sense as:
`"Reloading %s configuration failed with error: %v"

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [inline pkg/daemon/update.go:133]: same as above..

"Error processing state ang configs: %v, err"

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [inline pkg/daemon/update.go:139]: "Setting node's state to Done failed with error: %v"

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [reply pkg/daemon/daemon.go:1130]: is there are case where you return false, nil? The "All good!" comment below implies no?

**Review by kikisdeliveryservice — COMMENTED** [review]
some questions/comments

File: pkg/daemon/daemon.go
Comment (yuqi-zhang) [reply pkg/daemon/daemon.go:1130]: There is one case where we get `false, nil`, if `inDesiredConfig := state.currentConfig.GetName() == state.desiredConfig.GetName()` evaluates as false, it skips the whole if block and just returns false.

Comment (yuqi-zhang) [reply pkg/daemon/daemon.go:1130]: The naming is somewhat confusing, we probably should have named it `isInDesiredConfig` or something originally. But `inDesiredConfig` can be false

File: pkg/daemon/update.go
Comment (yuqi-zhang) [reply pkg/daemon/update.go:111]: Hmm, as in, add events for non-reboot scenarios? (We do have reboot events I believe)

File: test/e2e/mcd_test.go
Comment (yuqi-zhang) [reply test/e2e/mcd_test.go:157]: Will move

File: pkg/daemon/update.go
Comment (yuqi-zhang) [reply pkg/daemon/update.go:122]: Hmm I'd like to highlight the fact that we need to reboot, how about something like: `Reloading %s configuration failed, node will reboot instead. Error: %v` ?

Comment (kikisdeliveryservice) [reply pkg/daemon/update.go:122]: that works too :smile:

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [reply pkg/daemon/update.go:111]: yes exactly wondering if they are covered sufficiently  (sorry for not being clear!)

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [reply pkg/daemon/daemon.go:1130]: could we delete that "All good!" comment below? since that's back from when it just returned nil?

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [reply pkg/daemon/update.go]: maybe postConfigChangeActionSkipReboot? tho i think the none also works

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [reply pkg/daemon/update.go:115]: (i can take or leave this was just wondering)

Comment (yuqi-zhang) [reply pkg/daemon/update.go:115]: Added some in the specific scenarios below:
`dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeNormal, "Reboot", "Config changes do not require reboot.")`
and
`dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeNormal, "Reboot", "Config changes do not require reboot. Service %s was reloaded.", serviceName)`

Should I also add some warnings for the failure cases? Or what would be a good way to handle those (they would get a reboot event regardless)

File: pkg/daemon/update.go
Comment (yuqi-zhang) [reply pkg/daemon/update.go]: I think e.g. reloading crio also "skips reboot", so I defaulted to "none" to mean "no extra action needed"

**Comment by kikisdeliveryservice** [comment]
Played around with this pre- jerry fixing up comments and it works fine for me. i tested that things work to not reboot but reload crio, etc... then made a mc that changed the ssh key and a file and it did indeed reboot.

I think rolling it out with no reboot was ~6min for 3 nodes vs 11 min to roll out for the same nodes with reboot so some time savings there!

Also the comment clarifications from the last push are really helpful for me and addressed my questions.

Will let runcom/walters take final looks but lgtm.

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [reply pkg/daemon/update.go]: good point, im fine with none :+1:

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [reply pkg/daemon/update.go:115]: we have events for failsdrain and fails to reconcile for ex:
`dn.logSystem("Reloading %s configuration failed, node will reboot instead. Error: %v", serviceName, err)`
maybe a similar one near?: `dn.logSystem("Reloading %s configuration failed, node will reboot instead. Error: %v", serviceName, err)`

Comment (yuqi-zhang) [reply pkg/daemon/update.go:115]: Did you mean this line: `dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeWarning, "FailedToDrain", failMsg)` ?
So basically for what I added, instead of `Normal` I added it as a warning instead?

Comment (kikisdeliveryservice) [reply pkg/daemon/update.go:115]: oh duh yes sorry for the bad copypasta, yes that is the example, and yeah i thought maaybe put a warning near the reloading failed logging you already have to show it failed

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [inline pkg/daemon/update.go:129]: event here is what i mean @yuqi-zhang :laughing:

File: pkg/daemon/update.go
Comment (sinnykumari) [inline pkg/daemon/update.go:122]: Shouldn't the reason of event here be "SkipReboot" instead of "Reboot"?

File: pkg/daemon/update.go
Comment (sinnykumari) [reply pkg/daemon/update.go:135]: added warning event

**Comment by sinnykumari** [comment]
flake due to VpcLimitExceeded
/retest

**Comment by sinnykumari** [comment]
Need final round of review and approval get it merged.
/assign @cgwalters @runcom

**Review by cgwalters — APPROVED** [review]
I gave this a medium-level review and I think it looks good, but there are a lot of subtleties here and it touches code paths that we don't necessarily have CI coverage.

On the flip side, I think this is pretty safe to land because it will only kick in in very specialized circumstances.  We're very unlikely to break any default installs or CI.

That said, I do have one request (I wouldn't call this a blocker but a very nice to have): Let's add a new `machineconfiguration.openshift.io/bootedConfig` annotation that is set the first time we do a "live" update.

If you look at the recent [rpm-ostree livefs work](https://github.com/coreos/rpm-ostree/pull/2293) which is quite analogous to this, if you type `rpm-ostree status` you can clearly see that a "live" update was applied, as distinct from the booted commit.

Now currently nothing in rpm-ostree tries to capture the full "live history", and I don't think we need to do that here either.  The goal is that it should be *immediately obvious* looking at a node if it's in a "live applied" state.  That will help us debug problems in the future.

To clarify...anyone else should feel free to add a LGTM as is, but hopefully you agree with the above and we can try to add a separate `bootedConfig` as a followup soon.

**Comment by cgwalters** [comment]
If you haven't tried `rpm-ostree ex livefs`, here's a demo in current Fedora CoreOS.  There's a strong analogy between:

- "booted deployment" == "currentConfig"
- "staged deployment" == "desiredCommit"

Then when a livefs is applied, there's a new "livefs commit" (which is the same as staged, but note that staged could change without live changing later, so they need to be separately tracked).

```
```
walters@toolbox /v/s/w/b/fcos> cosa run
...
[root@cosa-devsh ~]# rpm-ostree install strace
...
Will download: 1 package (1.2?MB)
...
Added:
  strace-5.9-1.fc33.x86_64
Run "systemctl reboot" to start a reboot
[root@cosa-devsh ~]# rpm-ostree status
State: idle
Deployments:
  ostree://fedora:fedora/x86_64/coreos/testing-devel
                   Version: 33.20201119.dev.0 (2020-11-19T22:04:26Z)
                BaseCommit: 6d5b62bd02745cffbc4f3b470196eed8207fffc674c6602c428f632c226e5f2e
              GPGSignature: (unsigned)
                      Diff: 1 added
           LayeredPackages: strace

* ostree://fedora:fedora/x86_64/coreos/testing-devel
                   Version: 33.20201119.dev.0 (2020-11-19T22:04:26Z)
                    Commit: 6d5b62bd02745cffbc4f3b470196eed8207fffc674c6602c428f632c226e5f2e
              GPGSignature: (unsigned)
[root@cosa-devsh ~]# rpm-ostree ex livefs
[root@cosa-devsh ~]# rpm-ostree status
State: idle
Deployments:
  ostree://fedora:fedora/x86_64/coreos/testing-devel
                   Version: 33.20201119.dev.0 (2020-11-19T22:04:26Z)
                BaseCommit: 6d5b62bd02745cffbc4f3b470196eed8207fffc674c6602c428f632c226e5f2e
                    Commit: 0c398101141bfbf6c3661dd0d237622b865e2b85966007960e05a71cc7fcf1df
              GPGSignature: (unsigned)
                      Diff: 1 added
           LayeredPackages: strace

* ostree://fedora:fedora/x86_64/coreos/testing-devel
                   Version: 33.20201119.dev.0 (2020-11-19T22:04:26Z)
              BootedCommit: 6d5b62bd02745cffbc4f3b470196eed8207fffc674c6602c428f632c226e5f2e
                LiveCommit: 0c398101141bfbf6c3661dd0d237622b865e2b85966007960e05a71cc7fcf1df
                  LiveDiff: 1 added
              GPGSignature: (unsigned)
                  Unlocked: transient
[root@cosa-devsh ~]# 
```

File: pkg/daemon/update.go
Comment (kikisdeliveryservice) [reply pkg/daemon/update.go:135]: perf!

**Comment by kikisdeliveryservice** [comment]
I think @cgwalters suggestion could be a good followup for debugability (is that a word?)

@runcom any final thoughts?

File: pkg/daemon/update.go
Comment (yuqi-zhang) [inline pkg/daemon/update.go:533]: So this does change the behaviour of how the forcefile works slightly, but I guess its probably for the better, since this way it stays on the system until an update happens. Off the top of my head I don't think this should cause a problem. So +1

**Comment by yuqi-zhang** [comment]
Ok I think this is good to go! Let's aim to improve documentation and debugability in follow up PRs. Thanks everyone for the reviews!

/retest

**Comment by kikisdeliveryservice** [comment]
let's do this.

/lgtm

**Comment by kikisdeliveryservice** [comment]
everything passed! but now... tide is gonna... retest :neutral_face:

**Comment by runcom** [comment]
> To clarify...anyone else should feel free to add a LGTM as is, but hopefully you agree with the above and we can try to add a separate bootedConfig as a followup soon.

That is indeed as @kikisdeliveryservice commented a great thing to have for debuggability - we'll still must-gather in the cases we're investigating our usual bugs but it'll definitely help direct the investigation the right way. We can definitely follow up!

/lgtm

**Comment by sinnykumari** [comment]
>  hopefully you agree with the above and we can try to add a separate `bootedConfig` as a followup soon.

Nice and useful suggestion, definitely we will discuss and work on adding it.

**Pull request merged**
